src/resolver/environment.cc

```diff
@@ -520,7 +521,22 @@ Environment::tryResolveGroup( const InstallDescriptors & group,
             "we thought this was an unreachable error" );
         }
     }
-  return failure;
+  if ( failureForNewInput.empty() )
```

Comment:
non-blocking recommendation:

I'd recommend using a `std::unordered_map` here instead of a `std::vector<std::pair<...>>`.

You could also condense those conditionals a bit with less nesting to make it a bit easier to read:
```
if ( ! failureForNewInput.empty() ) { return failureForNewInput; }
if ( ! failureForOldInput.empty() ) { return failureForOldInput; }
throw RwsolutionFailureException( "..." );
```